### PR TITLE
NewtonsoftJsonInputFormatter checks for ErrorContext.Handled flag (#37323)

### LINKED ISSUE #37323: NewtonsoftJsonInputFormatter has incorrect behavior in case of user-added EventHandler
### Describe the bug
I want to add global ErrorHandler for Newtonsoft deserializer(as described at https://www.newtonsoft.com/json/help/html/SerializationErrorHandling.htm) . Purpose of my error handler is to ignore some exceptions, which are raised during deserialization of partially incorrect json. Simply put, I want to skip some wrong fields in json and leave their values in POCO uninitialized (default). This can be achieved with following code:

```
new JsonSerializerSettings
{
    Error = (sender, eventArgs) =>
    {
        if (eventArgs.ErrorContext.Error is JsonSerializationException ex &&
            ex.InnerException is InvalidCastException)
            eventArgs.ErrorContext.Handled = true;
}
```

This error handler works as intended in case of standalone usage with `JsonConvert.DeserializeObject`, but if I add the same handler to Asp.Net Core Startup at AddNewtonsoftJson, it doesn't work. This happens because ErrorHandler, which is  attached by Asp.Net at [NewtonsoftJsonInputFormatter  ](https://github.com/dotnet/aspnetcore/blob/d39e1ed64ebb4a1f6e921a1f1f1c88b02f45904f/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L233) does not respect `ErrorContext.Handled`. I think it is an error on Asp.Net part, and it's internal ErrorHandler should check this flag and exit from handler if it is already set to true.

### Further technical details
- ASP.NET Core 5
- Include the output of `dotnet --info`:

<details>
  <summary>dotnet --info Output</summary>

```
Пакет SDK для .NET (отражающий любой global.json):
 Version:   6.0.100-rc.1.21463.6
 Commit:    e627d556a1

Среда выполнения:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rc.1.21463.6\

Host (useful for support):
  Version: 6.0.0-rc.1.21451.13
  Commit:  d7619cd4b1

.NET SDKs installed:
  2.1.525 [C:\Program Files\dotnet\sdk]
  3.1.402 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.400 [C:\Program Files\dotnet\sdk]
  5.0.401 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.1.21463.6 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.1.21452.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

</details>

### PR DESCRIPTION
**Pull request by gbtb**
# NewtonsoftJsonInputFormatter checks for ErrorContext.Handled flag

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Added check for eventArgs.ErrorContext.Handled flag before reporting error.

Fixes #37323

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39275) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39275) before we can accept your contribution.<br/><br/>:x: gbtb [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=39275)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=39275) it.</sub>

File: src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs [hunk 1]
```diff
@@ -232,6 +232,11 @@ public virtual InputFormatterExceptionPolicy ExceptionPolicy
 
         void ErrorHandler(object? sender, Newtonsoft.Json.Serialization.ErrorEventArgs eventArgs)
         {
+            // Skipping error, if it's already marked as handled
+            // This allows user code to implement its own error handling
+            if (eventArgs.ErrorContext.Handled)
+                return;
```
Comment (TanayParikh) [inline src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs:238]: Nit;

```suggestion
            if (eventArgs.ErrorContext.Handled)
            {
                return;
            }
```

File: src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs [hunk 2]
```diff
@@ -487,6 +487,45 @@ public async Task ReadAsync_RegistersFileStreamForDisposal()
...
+            }
+        };
+        var formatter = new NewtonsoftJsonInputFormatter(
+            GetLogger(),
+            serializerSettings,
+            ArrayPool<char>.Shared,
+            _objectPoolProvider,
+            new MvcOptions(),
+            new MvcNewtonsoftJsonOptions());
+
```
Comment (pranavkm) [inline src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs:508]:
```suggestion
```

**Review by pranavkm — APPROVED** [review]
Thanks for the PR!

**Pull request merged**